### ISSUE TITLE
dobs plugin not handling underscores from Docker Compose (for Swarm)

### ISSUE THREAD

**Issue by oppodeldoc** [codedellemc/rexray#997]
# Summary
I was able to find some earlier issues referencing this and a [commit on libstorage](https://github.com/codedellemc/libstorage/blob/93f388ee10ae1f7c0fd6cb2f2779f8e4b635c6b0/drivers/storage/dobs/dobs.go#L62). Digital Ocean doesn't accept underscores in the names of volumes, so essentially the rexray/dobs plugin is useless in a docker stack using a compose file because `docker stack deploy` automatically prepends the project name and then an underscore to the volumes without any way to override it. 

# Bug Reports
I'm listing this as a bug because it looks like this was closed as part of #349 but now issues are gone from the libstorage repo so I can't see if this is now present as part of Rexray. Am I missing something, are there plans to fix this within this library or is it up to some other platform/vendor change? 

## Version
`$ docker plugin ls` returns:
```bash
ID                  NAME                 DESCRIPTION                               ENABLED
90ce57c1ce25        rexray/dobs:latest   REX-Ray for Digital Ocean Block Storage   true
```
`$ docker plugin inspect rexray/dobs` returns:
```json
[
    {
        "Config": {

            "Args": {
                "Description": "",
                "Name": "",
                "Settable": null,
                "Value": null
            },
            "Description": "REX-Ray for Digital Ocean Block Storage",
            "DockerVersion": "17.06.0-ce",
            "Documentation": "https://github.com/codedellemc/rexray/.docker/plugin/dobs",
            "Entrypoint": [
                "/rexray.sh",
                "rexray",
                "start",
                "-f",
                "--nopid"
            ],
            "Env": [
                {
                    "Description": "",
                    "Name": "REXRAY_FSTYPE",
                    "Settable": [
                        "value"
                    ],
                    "Value": "ext4"
                },
                {
                    "Description": "",
                    "Name": "REXRAY_LOGLEVEL",
                    "Settable": [
                        "value"
                    ],
                    "Value": "warn"
                },

                {
                    "Description": "",
                    "Name": "REXRAY_PREEMPT",
                    "Settable": [
                        "value"
                    ],
                    "Value": "false"
                },
                {
                    "Description": "",
                    "Name": "DOBS_REGION",
                    "Settable": [
                        "value"
                    ],
                    "Value": ""
                },
                {
                    "Description": "",
                    "Name": "DOBS_STATUSINITIALDELAY",
                    "Settable": [
                        "value"
                    ],
                    "Value": ""
                },
                {
                    "Description": "",
                    "Name": "DOBS_STATUSMAXATTEMPTS",
                    "Settable": [
                        "value"
                    ],
                    "Value": ""
                },
                {
                    "Description": "",
                    "Name": "DOBS_STATUSTIMEOUT",
                    "Settable": [
                        "value"
                    ],
                    "Value": ""
                },
                {
                    "Description": "",
                    "Name": "DOBS_TOKEN",
                    "Settable": [
                        "value"

                    ],
                    "Value": ""
                }
            ],
            "Interface": {
                "Socket": "rexray.sock",
                "Types": [
                    "docker.volumedriver/1.0"
                ]
            },
            "IpcHost": false,
            "Linux": {
                "AllowAllDevices": true,
                "Capabilities": [
                    "CAP_SYS_ADMIN"
                ],
                "Devices": null
            },
            "Mounts": [
                {
                    "Description": "",
                    "Destination": "/dev",
                    "Name": "",
                    "Options": [
                        "rbind"
                    ],
                    "Settable": null,
                    "Source": "/dev",
                    "Type": "bind"
                }
            ],
            "Network": {
                "Type": "host"
            },
            "PidHost": false,
            "PropagatedMount": "/var/lib/libstorage/volumes",
            "User": {},
            "WorkDir": "",
            "rootfs": {
                "diff_ids": [
                    "sha256:cf896aea29eb3156e8a5efe6bbb5b2dbf3322b3b634897db4b336fb0860ba647"
                ],
                "type": "layers"

            }
        },
        "Enabled": true,
        "Id": "90ce57c1ce25da092d1b89da628c8b2178247c04820300cb9b0da5b72dfd9e2c",
        "Name": "rexray/dobs:latest",
        "PluginReference": "docker.io/rexray/dobs:latest",
        "Settings": {
            "Args": [],
            "Devices": [],
            "Env": [
                "REXRAY_FSTYPE=ext4",
                "REXRAY_LOGLEVEL=warn",
                "REXRAY_PREEMPT=false",
                "DOBS_REGION=sfo2",
                "DOBS_STATUSINITIALDELAY=",
                "DOBS_STATUSMAXATTEMPTS=",
                "DOBS_STATUSTIMEOUT=",
                "DOBS_TOKEN=XXXXXXXXXXX"
            ],
            "Mounts": [
                {
                    "Description": "",
                    "Destination": "/dev",
                    "Name": "",
                    "Options": [
                        "rbind"
                    ],
                    "Settable": null,
                    "Source": "/dev",
                    "Type": "bind"
                }
            ]
        }
    }
]
```
## Expected Behavior
Should be possible to set `DOBS_CONVERTUNDERSCORES` during plugin setup or pass in an option to the `volumes` object in `docker-compose.yml`

## Actual Behavior
I found an issue referencing a variable `DOBS_CONVERTUNDERSCORES`, so I tried:
`$ docker set rexray/dobs CONVERTUNDERSCORES=true`
and got
`Error response from daemon: setting "DOBS_CONVERTUNDERSCORES" not found in the plugin configuration`

## Steps To Reproduce
As above, under Actual Behavior, I tried setting this option during and after install from the command line. I also tried passing in an option in `docker-compose.yml` like this:
```yaml
volumes:
  mysqldata:
    driver: rexray/dobs
    driver_opts:
      convertUnderscores: true
```
To no avail. 

## Logs
The error I'm getting back from the Digital Ocean API looks like this:
```bash
"create testing_mysqldata: VolumeDriver.Create: {"Error":"POST https://api.digitalocean.com/v2/volumes: 422 (request "7ffa5f4a-f3d5-48e2-97ae-fc755262ac53") Name must start with a lowercase letter followed by up to 62 lowercase letters, numbers, or hyphens, and cannot end with a hyphen"}
```

Am I missing some magic setting here? I was able to find this pretty well documented in the [libstorage docs](https://libstorage.readthedocs.io/en/stable/user-guide/storage-providers/#digitalocean), but it's gone in RexRay. Was this removed for any specific reason?

Any help or workarounds would be great. Thanks for all the hard work on this plugin!

**Comment by patrickleet** [comment]
it goes in your libstorage config

```
libstorage:
  service: dobs
  server:
    services:
      dobs:
        driver: dobs
        dobs:
          token: ${dobs_token}
          region: ${swarm_region}
          tag: rexray
          convertUnderscores: true
          statusMaxAttempts: 10
          statusInitialDelay: 100ms

```

**Comment by patrickleet** [comment]
that goes in `/etc/rexray/config.yml` fyi

**Comment by oppodeldoc** [comment]
Hi @patrickleet thanks for your response. I'm confused here because I thought configs were only used in standalone mode. I'm using this as a plugin, installed with `docker plugin install`. I tried adding this config to all nodes in my swarm but it doesn't affect anything.

**Comment by codenrhoden** [comment]
@oppodeldoc It does not appear that we have exposed that config option to the Docker containers. No good reason -- it just hasn't been added. The allowed env vars for the Docker plugins have to be explicitly listed out in a config file. For Digital Ocean, that is here: https://github.com/codedellemc/rexray/blob/master/.docker/plugins/dobs/config.json

So yes, I would not expect it to work right now. We'll need to add that option to the config file.

**Comment by oppodeldoc** [comment]
Ok, @codenrhoden thanks for confirming, I suspected that might be the case. Is this something pretty straightforward to add? I might have time to submit a PR in the next week unless someone on your side is on it already.

**Comment by codenrhoden** [comment]
@oppodeldoc Yes, fortunately it is that simple. Just add:
```json
        {
          "Description": "",
          "Name": "DOBS_CONVERTUNDERSCORES",
          "Settable": [
            "value"
          ],
          "Value": "false"
        }
```
to the file I linked to before. That should do it.

If you can put a PR together for it, that would be fantastic. We are gearing up for a new release next week that has focused on items not related to storage drivers and docker plugins, but those will get a renewed focus right after. Having this queued up for that would be great. And if it makes it in time, could make the next release as well.

**Comment by oppodeldoc** [comment]
Great, I'll go for it, thanks!

**Comment by akutz** [comment]
Closed with #1004.
